File: simulate_deals.js
function simulateDeals (N, upfrontCost, paybackMonths, attritionRate, 
                        renewalRate, contractLength, months)
{
        var totalUpfront = N * upfrontCost;
        var fee = upfrontCost / paybackMonths;
        var revenueStream = [];
        for (var i = 0; i < months; i++) { revenueStream.push(0); }

        var individualResults = [];
        for (var i = 0; i < N; i++)
        {
                var feeVector = contractStream(fee, attritionRate, renewalRate, contractLength, months);
                individualResults.push(feeVector);
                for (var j = 0; j < months; j++)
                {
                        revenueStream[j] += feeVector[j];
                }
        }

        var totalRevenue = 0;
        for (var i = 0; i < months; i++)
        {
                totalRevenue += revenueStream[i];
        }

        var totalVector = revenueStream; // new variable to calculate IRR
        totalVector.unshift(-totalUpfront); // need to add upfront cost
        var returns = irr(totalVector);
        returns = Math.pow((1 + returns), 12) - 1; // convert monthly IRR to annual

        var result =
        {
                capex: totalUpfront,
                irr: returns,
                stream: revenueStream,
                totalRevenue: totalRevenue,
                projects: individualResults
        };

        return result;
}

File: rb/simulate_deals.rb
module Projection
  def simulate_deals(n, upfront_cost, payback_months, attrition_rate,
                     renewal_rate, contract_length, months)
    total_upfront = n * upfront_cost
    fee = upfront_cost / payback_months
    revenue_stream = []
    months.times do revenue_stream << 0 end

    n.times do |deal|
      fee_vector = contract_stream(fee, attrition_rate, renewal_rate,
                                   contract_length, months)
      fee_vector.each_with_index do |fee, index|
        revenue_stream[index] += fee
      end
    end

    result = { :capex => total_upfront, :revenues => revenue_stream }
    result
  end
end
